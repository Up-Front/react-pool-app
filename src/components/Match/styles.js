import styled from 'styled-components';

export const MatchWrapper = styled.li`
  position: relative;
  display: flex;
  flex-direction: row;
  align-items: center;
  padding: 10px 0;
  width: 100%;
  border-bottom: 1px solid ${props => props.theme.borderColor};

  background-color: ${props =>
    props.contested
      ? props.theme.contestedMatchBackgroundColor
      : props.theme.matchBackgroundColor};
  color: ${props => props.theme.fontColor};
`;

export const Versus = styled.div`
  width: 50px;
  text-align: center;
  font-weight: bold;
  font-size: 24px;
`;

export const Head2Head = styled.div`
  font-weight: normal;
  font-size: 16px;
`;
