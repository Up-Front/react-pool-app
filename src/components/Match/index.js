import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Constants from './../shared/constants';
import Competitor from './components/Competitor';
import SwipeDelete from './../shared/components/SwipeDelete';
import { declareWinner, removeMatch } from './../../actions/matches';
import { MatchWrapper, Versus, Head2Head } from './styles';

class Match extends Component {
  /**
   * user can only declare winner, if he is 1 of the competitors
   * and only 1 time
   */
  handleDeclareWinner = competitor => {
    declareWinner(
      this.props.matchId,
      this.props.match,
      competitor,
      this.props.auth
    );
  };

  /**
   * check if the auth has voted for the competitor
   * so we can show that on the component
   * the votes are in the match.winners property
   */
  checkAuthVote(competitor) {
    if (
      this.props.match.winners &&
      this.props.match.winners[this.props.auth.uid] &&
      this.props.match.winners[this.props.auth.uid] === competitor.uid
    ) {
      return true;
    }
    return false;
  }

  /**
   * check if the authUser is a competitor in this match
   */
  checkAuthIsCompetitor(competitors) {
    return !!Object.values(competitors)
      .filter(competitor => competitor.uid === this.props.auth.uid)
      .shift();
  }

  handleRemoveMatch(matchId, match) {
    removeMatch(matchId, match)
      .then(() => {
        console.log('match removed');
      })
      .catch(() => {
        console.log('something went oops');
      });
  }

  canBeDeleted(match) {
    return (
      !this.hasWinner(match) && this.checkAuthIsCompetitor(match.competitors)
    );
  }

  hasWinner(match) {
    return !!match.winner;
  }

  //find the head2head score of the 2 competitors
  calcHead2Head(competitorA, competitorB) {
    const streak = competitorA.results && competitorA.results[competitorB.uid];
    let wins = 0;
    let losses = 0;
    if (streak) {
      wins = (streak.match(new RegExp(Constants.winValue, 'g')) || []).length;
      losses = (streak.match(new RegExp(Constants.loseValue, 'g')) || [])
        .length;
    }
    return `${wins}-${losses}`;
  }

  renderCompetitor(competitor, align) {
    return (
      <Competitor
        align={align}
        key={competitor.uid}
        checkAuthIsCompetitor={this.checkAuthIsCompetitor(
          this.props.match.competitors
        )}
        competitor={competitor}
        hasVote={this.checkAuthVote(competitor)}
        winner={this.props.match.winner}
        handleClick={this.handleDeclareWinner}
      />
    );
  }

  renderMatch() {
    const contestedText = this.props.isContested
      ? 'this match result is contested'
      : '';

    const competitorA = this.props.match.competitors[0];
    const competitorB = this.props.match.competitors[1];

    return (
      <MatchWrapper contested={this.props.match.isContested}>
        <strong>{contestedText}</strong>
        {this.renderCompetitor(competitorA, Constants.ALIGNLEFT)}
        <Versus>
          vs
          <Head2Head>{this.calcHead2Head(competitorA, competitorB)}</Head2Head>
        </Versus>
        {this.renderCompetitor(competitorB, Constants.ALIGNRIGHT)}
      </MatchWrapper>
    );
  }

  render() {
    if (this.canBeDeleted(this.props.match)) {
      return (
        <SwipeDelete
          key={this.props.matchId}
          deleteId={this.props.matchId}
          deleteObject={this.props.match}
          canbeSwiped={this.canBeDeleted(this.props.match)}
          onDelete={this.handleRemoveMatch}
        >
          {this.renderMatch()}
        </SwipeDelete>
      );
    } else {
      return this.renderMatch();
    }
  }
}

Match.propTypes = {
  matchId: PropTypes.string.isRequired,
  match: PropTypes.object.isRequired,
  auth: PropTypes.object.isRequired,
};

export default Match;
