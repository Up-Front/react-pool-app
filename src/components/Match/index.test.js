import React from 'react';
import { shallow, mount } from 'enzyme';
import Match from './';


describe('Match component', () => {
    let match;
    let auth;
    let matchId;
    beforeEach(() => {
        matchId = '1';
        auth = {
            uid: 1,
            displayName: 'Bob de Bouwer'
        };
        match = {
            uid: '1',
            competitors: {
                0: {
                    uid: 1,
                    displayName: 'Bob de Bouwer',
                },
                1: {
                    uid: 2,
                    displayName: 'Asterix',
                }
            },
        };
    });

    test('renders without crashing', () => {
        const wrapper = shallow(<Match
            matchId={matchId}
            auth={auth}
            match={match}
        />);

        expect(wrapper.find('.deletebutton').length).toEqual(1);
        expect(wrapper).toMatchSnapshot();
    });

    test('when there is a winner, there is no delete button', () => {

        match = {
            uid: '1',
            winner: '2',
            competitors: {
                0: {
                    uid: 1,
                    displayName: 'Bob de Bouwer',
                },
                1: {
                    uid: 2,
                    displayName: 'Asterix',
                }
            },
        };

        const wrapper = shallow(<Match
            matchId={matchId}
            auth={auth}
            match={match}
        />);

        const result = wrapper.instance().checkForWinner(match);
        expect(result).toEqual(true);

        expect(wrapper.find('.deletebutton').length).toEqual(0);

    });

    test('remove click', () => {
        const wrapper = mount(<Match
            matchId={matchId}
            auth={auth}
            match={match}
        />);

        wrapper.instance().handleRemoveMatch = jest.fn();
        wrapper.instance().forceUpdate();
        wrapper.find('.deletebutton').simulate('click');
        expect(wrapper.instance().handleRemoveMatch).toHaveBeenCalledTimes(1);
    });


    test('check that auth has voted', () => {
        match = {
            uid: '1',
            competitors: {
                0: {
                    uid: 1,
                    displayName: 'Bob de Bouwer',
                },
                1: {
                    uid: 2,
                    displayName: 'Asterix',
                }
            },
            winners: {
                '1': 2
            }
        };

        const wrapper = shallow(<Match
            matchId={matchId}
            auth={auth}
            match={match}
        />);

        const result = wrapper.instance().checkAuthVote(match.competitors[1]);
        expect(result).toEqual(true);
    });



});