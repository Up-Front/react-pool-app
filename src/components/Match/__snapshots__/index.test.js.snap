// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Match component renders without crashing 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Match
    auth={
        Object {
            "displayName": "Bob de Bouwer",
            "uid": 1,
          }
    }
    currentRanking={Object {}}
    match={
        Object {
            "competitors": Object {
              "0": Object {
                "displayName": "Bob de Bouwer",
                "uid": 1,
              },
              "1": Object {
                "displayName": "Asterix",
                "uid": 2,
              },
            },
            "uid": "1",
          }
    }
    matchId="1"
    previousRanking={Object {}}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": "1",
    "nodeType": "class",
    "props": Object {
      "canbeSwiped": true,
      "children": <styled.li
        contested={undefined}
>
        <strong />
        <Competitor
                align="left"
                checkAuthIsCompetitor={true}
                competitor={
                        Object {
                                "displayName": "Bob de Bouwer",
                                "uid": 1,
                              }
                }
                currentRanking={Object {}}
                handleClick={[Function]}
                hasVote={false}
                online={false}
                previousRanking={Object {}}
                winner={undefined}
        />
        <styled.div>
                vs
                <styled.div>
                        0-0
                </styled.div>
        </styled.div>
        <Competitor
                align="right"
                checkAuthIsCompetitor={true}
                competitor={
                        Object {
                                "displayName": "Asterix",
                                "uid": 2,
                              }
                }
                currentRanking={Object {}}
                handleClick={[Function]}
                hasVote={false}
                online={false}
                previousRanking={Object {}}
                winner={undefined}
        />
</styled.li>,
      "deleteId": "1",
      "deleteObject": Object {
        "competitors": Object {
          "0": Object {
            "displayName": "Bob de Bouwer",
            "uid": 1,
          },
          "1": Object {
            "displayName": "Asterix",
            "uid": 2,
          },
        },
        "uid": "1",
      },
      "onDelete": [Function],
      "tag": "div",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <strong />,
          <Competitor
            align="left"
            checkAuthIsCompetitor={true}
            competitor={
                        Object {
                                    "displayName": "Bob de Bouwer",
                                    "uid": 1,
                                  }
            }
            currentRanking={Object {}}
            handleClick={[Function]}
            hasVote={false}
            online={false}
            previousRanking={Object {}}
            winner={undefined}
/>,
          <styled.div>
            vs
            <styled.div>
                        0-0
            </styled.div>
</styled.div>,
          <Competitor
            align="right"
            checkAuthIsCompetitor={true}
            competitor={
                        Object {
                                    "displayName": "Asterix",
                                    "uid": 2,
                                  }
            }
            currentRanking={Object {}}
            handleClick={[Function]}
            hasVote={false}
            online={false}
            previousRanking={Object {}}
            winner={undefined}
/>,
        ],
        "contested": undefined,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "",
          },
          "ref": null,
          "rendered": "",
          "type": "strong",
        },
        Object {
          "instance": null,
          "key": "1",
          "nodeType": "function",
          "props": Object {
            "align": "left",
            "checkAuthIsCompetitor": true,
            "competitor": Object {
              "displayName": "Bob de Bouwer",
              "uid": 1,
            },
            "currentRanking": Object {},
            "handleClick": [Function],
            "hasVote": false,
            "online": false,
            "previousRanking": Object {},
            "winner": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              "vs",
              <styled.div>
                0-0
</styled.div>,
            ],
          },
          "ref": null,
          "rendered": Array [
            "vs",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "0-0",
              },
              "ref": null,
              "rendered": "0-0",
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "2",
          "nodeType": "function",
          "props": Object {
            "align": "right",
            "checkAuthIsCompetitor": true,
            "competitor": Object {
              "displayName": "Asterix",
              "uid": 2,
            },
            "currentRanking": Object {},
            "handleClick": [Function],
            "hasVote": false,
            "online": false,
            "previousRanking": Object {},
            "winner": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": "1",
      "nodeType": "class",
      "props": Object {
        "canbeSwiped": true,
        "children": <styled.li
          contested={undefined}
>
          <strong />
          <Competitor
                    align="left"
                    checkAuthIsCompetitor={true}
                    competitor={
                              Object {
                                        "displayName": "Bob de Bouwer",
                                        "uid": 1,
                                      }
                    }
                    currentRanking={Object {}}
                    handleClick={[Function]}
                    hasVote={false}
                    online={false}
                    previousRanking={Object {}}
                    winner={undefined}
          />
          <styled.div>
                    vs
                    <styled.div>
                              0-0
                    </styled.div>
          </styled.div>
          <Competitor
                    align="right"
                    checkAuthIsCompetitor={true}
                    competitor={
                              Object {
                                        "displayName": "Asterix",
                                        "uid": 2,
                                      }
                    }
                    currentRanking={Object {}}
                    handleClick={[Function]}
                    hasVote={false}
                    online={false}
                    previousRanking={Object {}}
                    winner={undefined}
          />
</styled.li>,
        "deleteId": "1",
        "deleteObject": Object {
          "competitors": Object {
            "0": Object {
              "displayName": "Bob de Bouwer",
              "uid": 1,
            },
            "1": Object {
              "displayName": "Asterix",
              "uid": 2,
            },
          },
          "uid": "1",
        },
        "onDelete": [Function],
        "tag": "div",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <strong />,
            <Competitor
              align="left"
              checkAuthIsCompetitor={true}
              competitor={
                            Object {
                                          "displayName": "Bob de Bouwer",
                                          "uid": 1,
                                        }
              }
              currentRanking={Object {}}
              handleClick={[Function]}
              hasVote={false}
              online={false}
              previousRanking={Object {}}
              winner={undefined}
/>,
            <styled.div>
              vs
              <styled.div>
                            0-0
              </styled.div>
</styled.div>,
            <Competitor
              align="right"
              checkAuthIsCompetitor={true}
              competitor={
                            Object {
                                          "displayName": "Asterix",
                                          "uid": 2,
                                        }
              }
              currentRanking={Object {}}
              handleClick={[Function]}
              hasVote={false}
              online={false}
              previousRanking={Object {}}
              winner={undefined}
/>,
          ],
          "contested": undefined,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "",
            },
            "ref": null,
            "rendered": "",
            "type": "strong",
          },
          Object {
            "instance": null,
            "key": "1",
            "nodeType": "function",
            "props": Object {
              "align": "left",
              "checkAuthIsCompetitor": true,
              "competitor": Object {
                "displayName": "Bob de Bouwer",
                "uid": 1,
              },
              "currentRanking": Object {},
              "handleClick": [Function],
              "hasVote": false,
              "online": false,
              "previousRanking": Object {},
              "winner": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": Array [
                "vs",
                <styled.div>
                  0-0
</styled.div>,
              ],
            },
            "ref": null,
            "rendered": Array [
              "vs",
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": "0-0",
                },
                "ref": null,
                "rendered": "0-0",
                "type": [Function],
              },
            ],
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "2",
            "nodeType": "function",
            "props": Object {
              "align": "right",
              "checkAuthIsCompetitor": true,
              "competitor": Object {
                "displayName": "Asterix",
                "uid": 2,
              },
              "currentRanking": Object {},
              "handleClick": [Function],
              "hasVote": false,
              "online": false,
              "previousRanking": Object {},
              "winner": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
